tests/scipy_optimize_test.py

```diff
@@ -0,0 +1,289 @@
+    c = 0
+    smax = 512
+    for name, f, fprime, x, p, old_f in self.line_iter():
+      f0 = f(x)
+      g0 = fprime(x)
+      self.fcount = 0
+      res = line_search(value_and_grad(f, fprime), x, p, f_0=f0, g_0=g0)
+      s = res.a_k
+      fc = res.nfev
+      gc = res.ngev
+      fv = res.f_k
+      gv = res.g_k
+      # s, fc, gc, fv, ofv, gv = ls.line_search_wolfe2(f, fprime, x, p,
+      #                                                g0, f0, old_f,
+      #                                                amax=smax)
+      # assert_equal(self.fcount, fc+gc)
+      self.assertAllClose(fv, f(x + s * p), check_dtypes=False, atol=1e-5)
+      if gv is not None:
+        self.assertAllClose(gv, fprime(x + s * p), check_dtypes=False, atol=1e-5)
+      if s < smax:
+        c += 1
+    self.assertTrue(c > 3)  # check that the iterator really works...
+
+  def test_line_search_wolfe2_bounds(self):
+    # See gh-7475
+
+    # For this f and p, starting at a point on axis 0, the strong Wolfe
+    # condition 2 is met if and only if the step length s satisfies
+    # |x + s| <= c2 * |x|
+    f = lambda x: jnp.dot(x, x)
+    fp = lambda x: 2 * x
+    p = jnp.array([1, 0])
+
+    # Smallest s satisfying strong Wolfe conditions for these arguments is 30
+    x = -60 * p
+    c2 = 0.5
+
+    res = line_search(value_and_grad(f, fp), x, p, c2=c2)
+    s = res.a_k
+    # s, _, _, _, _, _ = ls.line_search_wolfe2(f, fp, x, p, amax=30, c2=c2)
+    self.assert_line_wolfe(x, p, s, f, fp)
+    self.assertTrue(s >= 30.)
+
+    res = line_search(value_and_grad(f, fp), x, p, c2=c2, max_iterations=5)
+    self.assertTrue(res.failed)
+    # s=30 will only be tried on the 6th iteration, so this won't converge
+
+  def test_line_search(self):
+    import jax
+
+    import jax.numpy as np
+
+    def f(x):
+      return np.cos(np.sum(np.exp(-x)) ** 2)
+
+    # assert not line_search(jax.value_and_grad(f), np.ones(2), np.array([-0.5, -0.25])).failed
+    xk = np.ones(2)
+    pk = np.array([-0.5, -0.25])
+    res = line_search(jax.value_and_grad(f), xk, pk, max_iterations=100)
+
+    from scipy.optimize.linesearch import line_search_wolfe2
+
+    scipy_res = line_search_wolfe2(f, jax.grad(f), xk, pk)
+
+    # print(scipy_res[0], res.a_k)
+    # print(scipy_res[3], res.f_k)
+
+    self.assertAllClose(scipy_res[0], res.a_k, atol=1e-5, check_dtypes=False)
+    self.assertAllClose(scipy_res[3], res.f_k, atol=1e-5, check_dtypes=False)
+
+  # -- More specific tests
+
+
+class TestBFGS(jtu.JaxTestCase):
+
+  @parameterized.named_parameters(jtu.cases_from_list(
+    {"testcase_name": "_maxiter={}".format(maxiter), "maxiter": maxiter}
+    for maxiter in [None]))
+  def test_minimize(self, maxiter):
+    # TODO: Test this on maxiter=1,3,etc. up to inf (None), but this requires implementing DCSRCH to exactly replicate
+    # BFGS line search
+
+    from scipy.optimize import minimize as smin
+    import numpy as onp
+
+    def compare(func, x0):
+      # @jax.jit
```

Comment:
One other question -- does BFSG work under `jit` compilation?

If so, it would be nice to add some tests for that.

Reply:
It does indeed work under jit (at remarkable speed compared to scipy). What is the best way to assert something is jittable?

Reply:
Try to use the `_CompileAndCheck` helper function on `JaxTestCase`. It uses `jit` but also includes some extra checks against cache misses in compilation.

Reply:
Done.